docs/design/enhance-pluggable-policy.md

```diff
@@ -0,0 +1,175 @@
+The existing Authorization interfaces are described here: [authorization.md](../authorization.md).  A couple additions will allow the development of an Authorizer that matches based on different rules than the existing implementation.
+
+### Request Attributes
+The existing authorizer.Attributes only has 5 attributes (user, groups, isReadOnly, kind, and namespace).  If we add more detailed verbs, content, and resource names, then Authorizer plugins will have the same level of information available to RESTStorage components in order to express more detailed policy.  The replacement excerpt is below.
+
+A request has 7 attributes that can be considered for authorization:
+  - user - the user-string which a user was authenticated as.  This is included in the Context.
+  - groups - the groups to which the user belongs.  This is included in the Context.
+  - verb - string describing the requestion action. Today we have: get, list, watch, create, update, and delete
+  - what kind of object is being accessed 
+    - applies only to the API endpoints, such as 
+        `/api/v1beta1/pods`.  For miscelaneous endpoints, like `/version`, the kind is the empty string.
+  - the namespace of the object being access, or the empty string if the endpoint does not support namespaced objects.  This is included in the Context.
+  - content - the runtime.Object being submitted. This only has content for creates and updates
+  - resourceName - the name of the resource being gotten (get-ed) or deleted.
+
+### Authorizer Interface
+The existing Authorizer interface is very simple, but there isn't a way to provide details about allows, denies, or failures.  The extended detail is useful for UIs that want to describe why certain actions are disallowed.  Not all Authorizers will want to provide that information, but for those that do having that capability is useful.  In addition, adding a `GetAllowedSubjects` method that returns back the users and groups that can perform a particular action makes it possible to answer questions like, "who can see resources in my namespace" (see [ResourceAccessReview](#ResourceAccessReview) further down).
+
+```
+// OLD
+type Authorizer interface {
+  Authorize(a Attributes) error 
+}
+```
+
+```
+// NEW
+type Authorizer interface {
+  // Authorize takes a Context (for namespace, user, and traceability) and Attributes to make a policy determination.
+  // reason is an optional return value that can describe why a policy decision was made.
+  Authorize(ctx api.Context, a Attributes) (allowed bool, reason string, evaluationError error)
+
+  // GetAllowedSubjects takes a Context (for namespace and traceability) and Attributes to determine which users and 
+  // groups are allowed to perform the described action in the namespace.    This API enables the ResourceBasedReview requests below
+  GetAllowedSubjects(ctx api.Context, a Attributes) (users util.StringSet, groups util.StringSet, evaluationError error)
```

Comment:
I don't think ResourceAccessReview is required for all deployments or
authorizers.  It is definitely useful, but not all consumers may choose to
invert the model (or even support inverting the model).

On Fri, Aug 7, 2015 at 11:38 AM, David Eads <notifications@github.com>
wrote:

> In docs/design/enhance-pluggable-policy.md
> <https://github.com/GoogleCloudPlatform/kubernetes/pull/12209#discussion_r36531053>
> :
>
> > +// OLD
> > +type Authorizer interface {
> > +  Authorize(a Attributes) error
> > +}
> > +```
> > +
> > +```
> > +// NEW
> > +type Authorizer interface {
> > +  // Authorize takes a Context (for namespace, user, and traceability) and Attributes to make a policy determination.
> > +  // reason is an optional return value that can describe why a policy decision was made.
> > +  Authorize(ctx api.Context, a Attributes) (allowed bool, reason string, evaluationError error)
> > +
> > +  // GetAllowedSubjects takes a Context (for namespace and traceability) and Attributes to determine which users and
> > +  // groups are allowed to perform the described action in the namespace.    This API enables the ResourceBasedReview requests below
> > +  GetAllowedSubjects(ctx api.Context, a Attributes) (users util.StringSet, groups util.StringSet, evaluationError error)
>
> Because you don't need audit? Unless every call site needs both methods
> (and they don't) I would organize these so that there are two fundamental
> interfaces, and if at the highest level you require both, require both.
>
> So would we make ResourceAccessReview optional? It's essential for doing
> things like "which namespaces/projects can I see".
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/GoogleCloudPlatform/kubernetes/pull/12209/files#r36531053>
> .
>



-- 
Clayton Coleman | Lead Engineer, OpenShift